### PR TITLE
[improve](cloud-mow) merge and remove old version of delete bitmap when cumulative compaction is done

### PR DESCRIPTION
**Pull request by hust-hhb**
When the count of delete bitmap is big, it may lead to high cpu use rate,  we can merge old version delete bitmap to reduce the count of delete bitmap when cumulative compaction is done to  reduce the use rate of cpu.
Here is an example:
before cu compaciton ,rowset layout is

    "cumulative point": 10,
    "rowsets": [
        "[0-1] 0 DATA NONOVERLAPPING 0200000000000000ffffffffffffffecffffffffffffffff 0",
        "[2-9] 1 DATA NONOVERLAPPING 0200000000002cc1a8447a4e0ba5850f773803ae2d534284 1.21 KB",
        "[10-10] 1 DATA NONOVERLAPPING 0200000000002d07a8447a4e0ba5850f773803ae2d534284 1.09 KB",
        "[11-11] 1 DATA NONOVERLAPPING 0200000000002d14a8447a4e0ba5850f773803ae2d534284 1.10 KB",
        "[12-12] 1 DATA NONOVERLAPPING 0200000000002d23a8447a4e0ba5850f773803ae2d534284 1.10 KB",
        "[13-13] 1 DATA NONOVERLAPPING 0200000000002d33a8447a4e0ba5850f773803ae2d534284 1.10 KB",
        "[14-14] 1 DATA NONOVERLAPPING 0200000000002d40a8447a4e0ba5850f773803ae2d534284 1.10 KB"
    ],
base rowset is [2-9], now cu compaciton input rowset range is [10-14], after compaciton, rowset layout is

    "cumulative point": 10,
    "rowsets": [
        "[0-1] 0 DATA NONOVERLAPPING 0200000000000000ffffffffffffffecffffffffffffffff 0",
        "[2-9] 1 DATA NONOVERLAPPING 0200000000002cc1a8447a4e0ba5850f773803ae2d534284 1.21 KB",
        "[10-14] 1 DATA NONOVERLAPPING 0200000000002d63a8447a4e0ba5850f773803ae2d534284 1.21 KB"
    ],
1. the delete bitmap of rowset [10-10] -> [11-11] -> [12-12] -> [13-13] -> [14-14] will be deleted when delete_expired_stale_rowsets
2. the delete bitmap of version 10~14, will do agg first, then update the agg result on verison 14 delete bitmap
3. if update sucess, add the delete bitmap of version 10~13 to the queue which name is to_remove_vec
4. when deleting delete bitmap on rowset  [10-10] -> [11-11] -> [12-12] -> [13-13] -> [14-14], the delete bitmap which mention on step 3 will be deleted too.

### CODE REVIEW

**Comment by hust-hhb** [comment]
run buildall

**Comment by hust-hhb** [comment]
run buildall

**Comment by hust-hhb** [comment]
run buildall

**Comment by hust-hhb** [comment]
run buildall

**Comment by hust-hhb** [comment]
run buildall

File: cloud/src/meta-service/meta_service.cpp [hunk 1]
```diff
@@ -2080,6 +2080,168 @@ void MetaServiceImpl::get_delete_bitmap_update_lock(google::protobuf::RpcControl
...
+        return;
+    }
+    auto table_id = request->table_id();
+    std::string lock_key = meta_delete_bitmap_update_lock_key({instance_id, table_id, -1});
+    LOG(INFO) << "(" << instance_id << ")"
+              << "start to remove delete bitmap lock initiator, table_id=" << table_id
+              << ", key=" << hex(lock_key) << ", initiator=" << request->initiator();
+    std::string lock_val;
+    err = txn->get(lock_key, &lock_val);
+    if (err != TxnErrorCode::TXN_OK) {
```
Comment (zhannngchen) [inline cloud/src/meta-service/meta_service.cpp:2173]: how about NOT_FOUND?

File: fe/fe-core/src/main/java/org/apache/doris/cloud/transaction/CloudGlobalTransactionMgr.java [hunk 2]
```diff
@@ -782,8 +784,8 @@ private void getDeleteBitmapUpdateLock(Map<Long, Set<Long>> tableToParttions, lo
                     LOG.warn("ignore get delete bitmap lock exception, transactionId={}, retryTime={}",
                             transactionId, retryTime, e);
                 }
-                // sleep random millis [20, 200] ms, avoid txn conflict
-                int randomMillis = 20 + (int) (Math.random() * (200 - 20));
+                // sleep random millis [20, 300] ms, avoid txn conflict
```
Comment (zhannngchen) [inline fe/fe-core/src/main/java/org/apache/doris/cloud/transaction/CloudGlobalTransactionMgr.java:787]: why changed it to 300?

File: be/src/cloud/cloud_cumulative_compaction.cpp [hunk 3]
```diff
@@ -340,9 +341,125 @@ Status CloudCumulativeCompaction::modify_rowsets() {
...
+                int64_t initiator = boost::hash_range(_uuid.begin(), _uuid.end()) &
+                                    std::numeric_limits<int64_t>::max();
+                DBUG_EXECUTE_IF("CloudCumulativeCompaction.modify_rowsets.get_mow_lock_failed", {
+                    get_st = Status::InternalError(
+                            "test fail to get_delete_bitmap_update_lock for "
+                            "tablet_id {}",
+                            cloud_tablet()->tablet_id());
+                });
+                if (get_st.ok()) {
+                    get_st = _engine.meta_mgr().get_delete_bitmap_update_lock(*cloud_tablet(), -1,
```
Comment (zhannngchen) [inline be/src/cloud/cloud_cumulative_compaction.cpp:401]: Is it need to add some retry? I think it's quite easy to fail to get the lock?

File: be/src/olap/tablet_meta.cpp [hunk 4]
```diff
@@ -1166,6 +1168,36 @@ void DeleteBitmap::merge(const DeleteBitmap& other) {
...
+    std::shared_lock l(stale_delete_bitmap_lock);
+    _stale_delete_bitmap.emplace(tuple, time_stamp);
+}
+
+void DeleteBitmap::remove_stale_delete_bitmap_from_queue() {
+    std::shared_lock l(stale_delete_bitmap_lock);
+    auto now = UnixMillis();
+    auto it = _stale_delete_bitmap.begin();
+    while (it != _stale_delete_bitmap.end()) {
+        if (now - it->second > config::tablet_rowset_stale_sweep_time_sec * 1000) {
```
Comment (zhannngchen) [inline be/src/olap/tablet_meta.cpp:1183]: the config `tablet_rowset_stale_sweep_time_sec` is not for cloud

File: be/src/cloud/cloud_cumulative_compaction.cpp (hunk 3 above)
Comment (zhannngchen) [reply be/src/cloud/cloud_cumulative_compaction.cpp:401]: as we discussed, the lock is unnecessary.

**Comment by hust-hhb** [comment]
run buildall

**Comment by hust-hhb** [comment]
run buildall

**Comment by hust-hhb** [comment]
run buildall

File: cloud/src/meta-service/meta_service.h [hunk 5]
```diff
@@ -631,6 +641,21 @@ class MetaServiceProxy final : public MetaService {
                   done);
     }
 
+    void remove_delete_bitmap(google::protobuf::RpcController* controller,
```
[the remark this answers is not in the document]
Comment (hust-hhb) [reply cloud/src/meta-service/meta_service.h:652]: update new delete bitmap will fali beacause of Fail to find method=doris.cloud.MetaService/update_delete_bitmap_without_lock, and this update action will retry 200 times, which will take about 1 min and finally lead to update new delete bitmap  fail, but compaciton finally will success  and do nothing on delete bitmap.

File: be/src/cloud/cloud_delete_bitmap_action.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    CloudTabletSPtr tablet = DORIS_TRY(_engine.tablet_mgr().get_tablet(tablet_id));
+    if (tablet == nullptr) {
+        return Status::NotFound("Tablet not found. tablet_id={}", tablet_id);
+    }
+
+    auto count = tablet->tablet_meta()->delete_bitmap().get_delete_bitmap_count();
+
+    rapidjson::Document root;
+    root.SetObject();
+    root.AddMember("delete_bitmap_count", count, root.GetAllocator());
```
Comment (zhannngchen) [inline be/src/cloud/cloud_delete_bitmap_action.cpp:99]: we'd better to add more statistics here?
such as total delete bitmap cardinality, delete bitmap size(after serialized).

File: cloud/src/meta-service/meta_service.cpp [hunk 7]
```diff
@@ -2080,6 +2080,154 @@ void MetaServiceImpl::get_delete_bitmap_update_lock(google::protobuf::RpcControl
...
+                return;
+            }
+            fdb_txn_size = 0;
+            TxnErrorCode err = txn_kv_->create_txn(&txn);
+            if (err != TxnErrorCode::TXN_OK) {
+                code = cast_as<ErrCategory::CREATE>(err);
+                msg = "failed to init txn";
+                return;
+            }
+            if (!check_delete_bitmap_lock(code, msg, ss, txn, instance_id, table_id,
```
Comment (zhannngchen) [inline cloud/src/meta-service/meta_service.cpp:2151]: don't need to check lock here?

File: be/src/cloud/cloud_delete_bitmap_action.cpp (hunk 6 above)
Comment (hust-hhb) [reply be/src/cloud/cloud_delete_bitmap_action.cpp:101]: That right, already remove this check

File: cloud/src/meta-service/meta_service.cpp (hunk 7 above)
Comment (hust-hhb) [reply cloud/src/meta-service/meta_service.cpp]: That right, already remove this check

**Comment by hust-hhb** [comment]
run buildall

**Comment by hust-hhb** [comment]
run buildall

File: be/src/olap/tablet_meta.cpp [hunk 8]
```diff
@@ -1089,6 +1091,14 @@ uint64_t DeleteBitmap::cardinality() const {
     return res;
 }
 
+size_t DeleteBitmap::get_size() const {
+    size_t charge = 0;
```
Comment (zhannngchen) [inline be/src/olap/tablet_meta.cpp:1095]: forgot to hold the read lock?

**Comment by hust-hhb** [comment]
run buildall

File: be/src/olap/tablet_meta.cpp (hunk 8 above)
Comment (hust-hhb) [reply be/src/olap/tablet_meta.cpp:1096]: done

**Review by zhannngchen — DISMISSED** [review]
LGTM

**Comment by hust-hhb** [comment]
run buildall

**Review by zhannngchen — DISMISSED** [review]
LGTM

File: cloud/src/meta-service/meta_service.cpp [hunk 9]
```diff
@@ -2080,6 +2080,146 @@ void MetaServiceImpl::get_delete_bitmap_update_lock(google::protobuf::RpcControl
...
+                                       request->versions(i), request->segment_ids(i)};
+        std::string key;
+        meta_delete_bitmap_key(key_info, &key);
+        delete_bitmap_keys.add_delete_bitmap_keys(key);
+    }
+
+    // remove old delete bitmap
+
+    // Update delete bitmap for curent txn
+    for (size_t i = 0; i < request->rowset_ids_size(); ++i) {
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service.cpp:2127]: the for loop may last more than 5 seconds which results in "txn too old" error

File: cloud/src/meta-service/meta_service.cpp [hunk 10]
```diff
@@ -2080,6 +2080,146 @@ void MetaServiceImpl::get_delete_bitmap_update_lock(google::protobuf::RpcControl
...
+            if (err != TxnErrorCode::TXN_OK) {
+                code = cast_as<ErrCategory::CREATE>(err);
+                msg = "failed to init txn";
+                return;
+            }
+        }
+        // splitting large values (>90*1000) into multiple KVs
+        cloud::put(txn.get(), key, val, 0);
+        fdb_txn_size = fdb_txn_size + key.size() + val.size();
+        LOG(INFO) << "xxx update delete bitmap put delete_bitmap_key=" << hex(key)
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service.cpp:2154]: there may be too many keys. 
print the first or the last one, an sumarized how many keys and size have been put.

File: cloud/src/meta-service/meta_service.cpp [hunk 11]
```diff
@@ -2080,6 +2080,146 @@ void MetaServiceImpl::get_delete_bitmap_update_lock(google::protobuf::RpcControl
...
+        return;
+    }
+
+    PendingDeleteBitmapPB delete_bitmap_keys;
+    for (size_t i = 0; i < request->rowset_ids_size(); ++i) {
+        MetaDeleteBitmapInfo key_info {instance_id, tablet_id, request->rowset_ids(i),
+                                       request->versions(i), request->segment_ids(i)};
+        std::string key;
+        meta_delete_bitmap_key(key_info, &key);
+        delete_bitmap_keys.add_delete_bitmap_keys(key);
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service.cpp:2121]:
```suggestion
        delete_bitmap_keys.add_delete_bitmap_keys(meta_delete_bitmap_key(key_info));
```

Comment (gavinchou) [reply cloud/src/meta-service/meta_service.cpp:2121]: why do we need this separate loop?

File: cloud/src/meta-service/meta_service.cpp [hunk 12]
```diff
@@ -2080,6 +2080,146 @@ void MetaServiceImpl::get_delete_bitmap_update_lock(google::protobuf::RpcControl
...
+        return;
+    }
+    std::unique_ptr<Transaction> txn;
+    TxnErrorCode err = txn_kv_->create_txn(&txn);
+    if (err != TxnErrorCode::TXN_OK) {
+        LOG(WARNING) << "failed to init txn";
+        return;
+    }
+    for (size_t i = 0; i < rowset_ids.size(); i++) {
+        auto delete_bitmap_start = meta_delete_bitmap_key(
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service.cpp:2208]: can we estimate how many keys in the range, if there are few of them. 
try to remove by keys.

And, we should summarize and log what have been removed, say, how many keys/ranges have been removed.

File: gensrc/proto/cloud.proto [hunk 13]
```diff
@@ -1550,6 +1562,8 @@ service MetaService {
     rpc update_delete_bitmap(UpdateDeleteBitmapRequest) returns(UpdateDeleteBitmapResponse);
     rpc get_delete_bitmap(GetDeleteBitmapRequest) returns(GetDeleteBitmapResponse);
     rpc get_delete_bitmap_update_lock(GetDeleteBitmapUpdateLockRequest) returns(GetDeleteBitmapUpdateLockResponse);
+    rpc update_delete_bitmap_without_lock(UpdateDeleteBitmapRequest) returns(UpdateDeleteBitmapResponse);
```
Comment (gavinchou) [inline gensrc/proto/cloud.proto:1565]: why we cannot use the existing rpc `rpc update_delete_bitmap(UpdateDeleteBitmapRequest)`?

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: be/src/cloud/cloud_cumulative_compaction.cpp [hunk 14]
```diff
@@ -340,9 +342,87 @@ Status CloudCumulativeCompaction::modify_rowsets() {
...
+                        *cloud_tablet(), new_delete_bitmap.get());
+            }
+            if (!update_st.ok()) {
+                std::stringstream ss;
+                ss << "failed to update delete bitmap for tablet=" << cloud_tablet()->tablet_id()
+                   << " st=" << update_st.to_string();
+                std::string msg = ss.str();
+                LOG(WARNING) << msg;
+            }
+            if (update_st.ok()) {
```
Comment (dataroaring) [inline be/src/cloud/cloud_cumulative_compaction.cpp:407]:
```suggestion
            } else {
```

File: be/src/olap/tablet_meta.cpp [hunk 15]
```diff
@@ -1166,6 +1178,49 @@ void DeleteBitmap::merge(const DeleteBitmap& other) {
...
+                auto start_bmk = std::get<1>(delete_bitmap_tuple);
+                auto end_bmk = std::get<2>(delete_bitmap_tuple);
+                remove(start_bmk, end_bmk);
+                to_delete.emplace_back(std::make_tuple(std::get<0>(start_bmk).to_string(), 0,
+                                                       std::get<2>(end_bmk)));
+            }
+            _stale_delete_bitmap.erase(version_str);
+        }
+    }
+    CloudStorageEngine& engine = ExecEnv::GetInstance()->storage_engine().to_cloud();
```
Comment (dataroaring) [inline be/src/olap/tablet_meta.cpp:1211]: What happens when tablet_id is -1 and to_delete is empty?

Comment (hust-hhb) [reply be/src/olap/tablet_meta.cpp:1211]: remove_old_version_delete_bitmap will do nothing on ms, but just return this function is a better way

File: cloud/src/meta-service/meta_service.cpp (hunk 12 above)
Comment (hust-hhb) [reply cloud/src/meta-service/meta_service.cpp:2208]: now can not estimate how many keys beacause the request param is provide segment nums, may be in next time, we can remove delete bitmap key in more accurate range. 
in meta_service_job.cpp, remove delete bitmap code is

            // remove delete bitmap of input rowset for MoW table
            if (compaction.has_delete_bitmap_lock_initiator()) {
                auto delete_bitmap_start =
                        meta_delete_bitmap_key({instance_id, tablet_id, rs.rowset_id_v2(), 0, 0});
                auto delete_bitmap_end = meta_delete_bitmap_key(
                        {instance_id, tablet_id, rs.rowset_id_v2(), INT64_MAX, INT64_MAX});
                txn->remove(delete_bitmap_start, delete_bitmap_end);
            }
and in get_delete_bitmap function, get delete bitmap code is
        MetaDeleteBitmapInfo start_key_info {instance_id, tablet_id, rowset_ids[i],
                                             begin_versions[i], 0};
        MetaDeleteBitmapInfo end_key_info {instance_id, tablet_id, rowset_ids[i], end_versions[i],
                                           INT64_MAX};
        std::string start_key;
        std::string end_key;
        meta_delete_bitmap_key(start_key_info, &start_key);
        meta_delete_bitmap_key(end_key_info, &end_key);
those key range is more bigger, we can make these key range more accurate in next time.

File: gensrc/proto/cloud.proto (hunk 13 above)
Comment (hust-hhb) [reply gensrc/proto/cloud.proto:1565]: update_delete_bitmap need to hold lock, this function no need to hold lock

**Comment by hust-hhb** [comment]
run buildall

**Comment by hust-hhb** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**